File: src/algorithms/mathandprob/GrayCode.java
package algorithms.mathandprob; 

import java.util.*; 

public class GrayCode {
    /* Function: Give a number of bits, return the sequence of gray code of that bits. 
     * Solution: Notice 0 bits of gray code is 0. 
     */ 
    
    public List<Integer> grayCode(int n) {
        List<Integer> res = new ArrayList<Integer>(); 
        if (n < 0) {
            return res; 
        }
        
        res.add(0); 
        int off = 1; 
        for (int i = 0; i < n; ++i) {
            int pt = res.size() - 1; 
            for (; pt >= 0; --pt) {
                res.add(res.get(pt) + off); 
            }
            off *= 2; 
        }
        
        return res; 
    }
}

File: wwwroot/js/gray-code.js
/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function(n) {
    if (n === 0) {
        return [];
    }

    let res = [0, 1];
    let mul = 1;
    for (let i = 1; i < n; i++) {
        mul *= 2;
        let l = res.length;
        for (let j = l - 1; j >= 0; j--) {
            res.push(mul + res[j]);
        }
    }

    return res;
};

document.getElementById("gray-code").innerHTML =
    grayCode(3);